cmd/osbuild-service-maintenance/aws.go

```diff
@@ -103,14 +163,164 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
 	return nil
 }
 
-func filterReservations(reservations []ec2types.Reservation) []string {
-	var instanceIDs []string
+func filterOnTooOld(instanceIDs []string, reservations []ec2types.Reservation) []string {
 	for _, res := range reservations {
 		for _, i := range res.Instances {
 			if i.LaunchTime.Before(time.Now().Add(-time.Hour * 2)) {
-				instanceIDs = append(instanceIDs, *i.InstanceId)
+				logrus.Infof("Instance %s is too old", *i.InstanceId)
+				if !slices.Contains(instanceIDs, *i.InstanceId) {
+					instanceIDs = append(instanceIDs, *i.InstanceId)
+				}
 			}
 		}
 	}
 	return instanceIDs
 }
+
+func getChildAndParentInstances(reservations []ec2types.Reservation) []struct {
+	Child  string
+	Parent string
+} {
+	var instanceData []struct {
+		Child  string
+		Parent string
+	}
+
+	for _, res := range reservations {
+		for _, i := range res.Instances {
+			for _, t := range i.Tags {
+				if *t.Key == "parent" {
+					instanceData = append(instanceData, struct {
+						Child  string
+						Parent string
+					}{
+						Child:  *i.InstanceId,
+						Parent: *t.Value,
+					})
+				}
+			}
+		}
+	}
+	return instanceData
+}
+
+func checkValidParent(childId string, parent []ec2types.Reservation) bool {
+	if len(parent) == 0 {
+		logrus.Infof("Instance %s has no parent, removing it", childId)
+		return false
+	}
+	if len(parent) != 1 {
+		logrus.Errorf("Instance %s has %d parents. That should never happen, not changing anything here.", childId, len(parent))
+		return true
+	}
+	if len(parent[0].Instances) == 0 {
+		logrus.Infof("Instance %s has no parent instance, removing it", childId)
+		return false
+	}
+	if len(parent[0].Instances) != 1 {
```

Comment:
not sure if all those checks are necessary?

Reply:
I tihnk it's ok